src/scf/potential_mixing.jl

```diff
@@ -0,0 +1,327 @@
+using Statistics
+
+# Quick and dirty Anderson implementation ... lacks important things like
+# control of the condition number of the anderson matrix.
+# Not particularly optimised. Also should be moved to NLSolve ...
+function AndersonAcceleration(m=Inf)
+    # Accelerates the iterative solution of f(V) = 0, where for our DFT case:
+    #    f(V) = Vext + Vhxc(ρ(V)) - V
+    # Further define
+    #    preconditioned update    Pf(V) = P⁻¹ f(V)
+    #    fixed-point map          g(V)  = V + α Pf(V)
+    # where the α may vary between steps.
+    #
+    # Finds the linear combination Vₙ₊₁ = g(Vₙ) + ∑ᵢ βᵢ (g(Vᵢ) - g(Vₙ))
+    # such that |Pf(Vₙ) + ∑ᵢ βᵢ (Pf(Vᵢ) - Pf(Vₙ))|² is minimal
+    #
+    Vs   = []  # The V     for each iteration
+    PfVs = []  # The Pf(V) for each iteration
+
+    function extrapolate(Vₙ, αₙ, PfVₙ)
+        # Gets the current Vₙ, Pf(Vₙ) and damping αₙ
+        #
+        Vₙ₊₁ = @. vec(Vₙ) + αₙ * vec(PfVₙ)
+        if !isempty(Vs)
+            M = hcat(PfVs...) .- vec(PfVₙ)  # Mᵢⱼ = (PfVⱼ)ᵢ - (PfVₙ)ᵢ
+            # We need to solve 0 = M' PfVₙ + M'M βs <=> βs = - (M'M)⁻¹ M' PfVₙ
+            βs = -M \ vec(PfVₙ)
+            @. for (iβ, β) in enumerate(βs)
+                Vₙ₊₁ += β .* (Vs[iβ] - vec(Vₙ) + αₙ * (PfVs[iβ] - vec(PfVₙ)))
+            end
+        end
+
+        push!(Vs,   vec(Vₙ))
+        push!(PfVs, vec(PfVₙ))
+        if length(Vs) > m
+            Vs   = Vs[2:end]
+            PfVs = PfVs[2:end]
+        end
+        @assert length(Vs) <= m
+
+        reshape(Vₙ₊₁, size(Vₙ))
+    end
+end
+
+
+function ScfAcceptStepImproving(;tol=1e-12, residual_ratio=0.9)
+    function accept_step(info, info_next)
+        # Accept if energy goes down or residual decreases
+        return (
+               info_next.energies.total - info.energies.total < tol
+            || norm(info_next.Pinv_δV) / norm(info.Pinv_δV) < residual_ratio
+        )
+    end
+end
+function ScfAcceptStepAll()
+    accept_step(info, info_next) = true
+end
+
+
+function scf_quadratic_model(info, info_next; modeltol=0.1)
+    T = eltype(ρ)
+
+    # Vout    = step(V), where step(V) = (Vext + Vhxc(ρ(Vin)))
+    # Vnext   = Vin + α0 * (Anderson(Vin, P⁻¹( Vout - Vin )) - Vin)
+    # ρin     = ρ(Vin)
+    # ρnext   = ρ(Vnext)
+    # α0 * δV = α0 * (Vnext - Vin) = α0 * (Anderson(Vin, P⁻¹( Vout - Vin )) - Vin)
+    #
+    Vin   = info.Vin
```

Comment:
In julia 1.7 you can simplify this to (; Vin, Vout) = info (although that's not that interesting here)

Reply:
Yes, I saw this.